clang/include/clang/Frontend/CompileJobCacheResult.h

```diff
@@ -0,0 +1,111 @@
+//===- CompileJobCacheResult.h ----------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_FRONTEND_COMPILEJOBCACHERESULT_H
+#define LLVM_CLANG_FRONTEND_COMPILEJOBCACHERESULT_H
+
+#include "clang/Basic/LLVM.h"
+#include "llvm/CAS/CASNodeSchema.h"
+#include "llvm/CAS/ObjectStore.h"
+
+namespace clang {
+namespace cas {
+class CompileJobResultSchema;
+
+class CompileJobCacheResult : public ObjectProxy {
+public:
+  /// Categorization for the output kinds that is used to decouple the
+  /// compilation cache key from the specific output paths.
+  enum class OutputKind : char {
+    Unknown,    ///< Unknown output kind. \c Output::Path must be set.
```

Comment:
The `Unknown` kind seems new. It is kind of scary to write a file to a path specified since if the path is not captured as part of the CacheKey, this path might not write to the correct location.

Reply:
Unknown kind preserves the existing behaviour for unknown output file paths. Previously this was implicit in how CASOutputBackend did remapping.

>  if the path is not captured as part of the CacheKey, this path might not write to the correct location.

This is true, but the default behaviour if there is an output we don't handle (e.g. newly added) is that it will be in the cache key.

Reply:
Make sense. I wonder if we should subclass in clang to make a new CASBackend that requires associate outputFile with a kind.

Reply:
I added a unit test since the "unknown" case doesn't seem to be covered by any lit tests.  I'd like feedback (also from @akyrtzi) on whether the "unknown" case should be an assertion failure.  I'm wary of this option since secondary output files are not enumerated anywhere in the compiler that will allow us to ensure we haven't missed something.  If someone adds a new secondary output and it shows up in the compiler invocation, it seems better to me for it to cause cache misses.  The only "dangerous" case would be if there is a secondary output that is somehow not in the compiler invocation.  I think the only way that would happen is if we created an implicit output that used the main output path, e.g. "foo.o.myNewOutput.txt".  But maybe a better way to fix that would be to canonicalize the compiler invocation during compilation instead of only canonicalizing it during cache key generation?

Reply:
I remember I found a lot of outputs generated using the main output file path with a slightly different suffix so that is a pretty high possibility.

Reply:
> on whether the "unknown" case should be an assertion failure

Would it work to not have "unknown" kind and fail the compilation if some path is not handled?

Reply:
Okay, I'll take a look at making it a hard error (report_fatal_error I guess, since it's not a user error) and see if I can find any unhandled cases across the compiler. A quick look showed "-time-trace" probably has this issue, for example.

For now I'll keep this specific to the caching build and reject unknown paths, but long term we should probably refactor OutputBackend to use an output kind instead of a path in createFile.